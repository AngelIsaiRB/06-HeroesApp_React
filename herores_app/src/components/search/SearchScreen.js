import React, { useMemo } from 'react'
import queryString from "query-string"
import { heroes } from '../../data/heroes';
import { HeroCard } from '../heores/HeroCard';
import { useForm } from '../hooks/useForm';
import {useLocation} from "react-router-dom";
import { getHeroesByName } from '../../selectors/getHeroesByName';

export const SearchScreen = ({history}) => {
    // para parsear la busqueda ejemplo: ?q=btman
    const location = useLocation();
    const {q=""} = queryString.parse(location.search);
    // 
    
    // hook personalizado de useForm
    const initialForm = {
        search: q,        
    };    
    const [ values, handleInputChange, reset ] = useForm( initialForm );
    // end hook personalizado
    const heroesFilter = useMemo(() => getHeroesByName(q), [q])
    // const heroesFilter = getHeroesByName(values.search);
    
    const handleSearch=(e)=>{
        // evita el recargar la pagina
        e.preventDefault();

        // para mandar un query param en la barra de busqueda
        history.push(`?q=${values.search}`)
    }

    return (
        <div>
            <h1>Search Screen</h1> 
            <hr/>
            <div className="row">
                <div className="col-5">
                    <h4> SearchForm</h4>
                    <hr/>
                    <form onSubmit={handleSearch}>
                        <input type="text"
                        onChange={handleInputChange}    
                        value={values.search}                    
                        name="search"
                        placeholder="Find your heroe"
                        className="form-control">
                        </input>
                        <button
                        className="btn m-1 btn-block btn-outline-primary"
                        type="submit"
                        >Serach</button>
                    </form>
                </div>

                <div className="col-7">
                    <h4>Results</h4>
                    <hr/>
                    {
                    (q === "") && <div className="alert alert-info">
                            Search a hero
                                 </div>                    
                    }
                    {
                    (q !== "" && heroesFilter.length === 0) && <div className="alert alert-danger">
                            hero not found {q}
                                 </div>                    
                    }

                    {
                        
                        heroesFilter.map(hero=> (
                            <HeroCard 
                            key={hero.id}
                            {...hero}
                            />
                        ))
                    }
                </div>
            </div>
        </div>
    )
}
