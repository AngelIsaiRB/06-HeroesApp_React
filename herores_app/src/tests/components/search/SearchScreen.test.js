import React from 'react';
import { shallow, mount } from 'enzyme'
import { MemoryRouter, Route } from 'react-router-dom';
import { SearchScreen } from '../../../components/search/SearchScreen';

describe('pruebas en <SearchScreen/>', () => {

    test('should de mostrase correctamente con valores por defecto ', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={["/search"]}>
                <Route path='/search' component={SearchScreen}>                    
                </Route>
            </MemoryRouter>
        );
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find(".alert-info").text().trim()).toBe("Search a hero");
    });

    test('should mostrar a batman y el input con el valor del query String', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={["/search?q=batman"]}>
                <Route path='/search' component={SearchScreen}>                    
                </Route>
            </MemoryRouter>
        );

        expect(wrapper.find("input").prop("value")).toBe("batman");
        expect(wrapper).toMatchSnapshot();
    });
    
    test('should de mostrar un error si no se encuentrael hero', () => {
        const heroError="batman123123"
        const wrapper = mount(
            <MemoryRouter initialEntries={[`/search?q=${heroError}`]}>
                <Route path='/search' component={SearchScreen}>                    
                </Route>
            </MemoryRouter>
        );
        expect(wrapper.find(".alert-danger").text().includes(heroError)).toBe(true);
        expect(wrapper).toMatchSnapshot();
    });

    test('should de llamar el push del history', () => {
       const history = {
           push: jest.fn(),
       } 
       const wrapper = mount(
           <MemoryRouter initialEntries={["/search?q=batman123"]}>
               <Route path='/search' component={()=><SearchScreen history={history}/>}>                   
               </Route>
           </MemoryRouter>
       );

       wrapper.find("input").simulate("change",{
            target:{
                name: "search", // del input del formulario cambia con cada formulario !!!!!!
                value:"batman"
            }
       });
       wrapper.find("form").prop("onSubmit")({preventDefault(){}});
       expect(history.push).toHaveBeenCalledWith(`?q=batman`);


    });
    
    

})