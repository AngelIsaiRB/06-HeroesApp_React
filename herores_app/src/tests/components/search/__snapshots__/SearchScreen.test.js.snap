// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en <SearchScreen/> should de mostrar un error si no se encuentrael hero 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
         SearchForm
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          className="form-control"
          name="search"
          onChange={[Function]}
          placeholder="Find your heroe"
          type="text"
          value="batman123123"
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Serach
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-danger"
      >
        hero not found 
        batman123123
      </div>
    </div>
  </div>
</div>
`;

exports[`pruebas en <SearchScreen/> should de mostrase correctamente con valores por defecto  1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
         SearchForm
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          className="form-control"
          name="search"
          onChange={[Function]}
          placeholder="Find your heroe"
          type="text"
          value=""
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Serach
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a hero
      </div>
    </div>
  </div>
</div>
`;

exports[`pruebas en <SearchScreen/> should mostrar a batman y el input con el valor del query String 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
         SearchForm
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          className="form-control"
          name="search"
          onChange={[Function]}
          placeholder="Find your heroe"
          type="text"
          value="batman"
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Serach
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div>
        <div
          className="card ms-3"
          style={
            Object {
              "maxWidth": 540,
            }
          }
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-md-4"
            >
              <img
                alt="Batman"
                className="card-img"
                src="./assets/heroes/dc-batman.jpg"
              />
            </div>
            <div
              className="col-md-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                   
                  Batman
                </h5>
                <p
                  className="card-text"
                >
                  Bruce Wayne
                   
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  >
                    Detective Comics #27
                  </small>
                </p>
              </div>
              <a
                href="/hero/dc-batman"
                onClick={[Function]}
              >
                Mas...
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
